joomla/joomla-cms
# [6.2] Multiload of article associations

### Summary of Changes

Load language associations for multiple articles and categories in a single query per each `ArticlesModel::getItems()` call.

These methods now can accept array of IDs and return results grouped by ID in this case.
If a single `int` is provided as ID, the results are returned as they were before, full BC.

`CategoriesHelper::getAssociations()`
`CategoryAssociationHelper::getCategoryAssociations()`
`AssociationHelper::getAssociations()`
`Associations::getAssociations`

Also, `CategoriesHelper::getAssociations()` is optimized and an extra query for each loaded category is removed by using `$advClause` param of `Associations::getAssociations()`.

`ArticlesModel::getItems()` now collects the article IDs to load the assocations for, next loads all associations in the single query.
Similar to I have already done with tags load.

### Testing Instructions

1. Install 5.0-beta2-dev
2. Install German language
3. Install Sample multingual data
4. Clone each English article "Article (en-gb)" (edit it and click Save As Copy 2 times), we have 3 en-GB articles now.
5. Clone each German article"Beitrag (de-de)" (edit it and click Save As Copy 2 times, we have 3 de-DE articles now.
6. Edit 2nd and 3rd copies of articles and associate them. Now we have 3 en-GB articles asociated with 3 de-DE articles.
7. Enable Joomla debug
8. Enable "Associations" in com_content options (they can be enabled in homepage menu item settings but it doesn't work for some reason, looks like another issuse, hence we need to enable display of associations globally).
9. Load frontend, see list of articles from a category with 37 SQL queries executed in debug bar, for each articles we load associated article in a separate query.
10. Apply patch, see a single query which loads associated articles for all displayed articles.

### Actual result BEFORE applying this Pull Request

37 queries in total, see a separate SQL query for each displayed article, like:

```
SELECT `c2`.`language`,CONCAT_WS(':', `c2`.`id`, `c2`.`alias`) AS `id`,CONCAT_WS(':', `ca`.`id`, `ca`.`alias`) AS `catid`
FROM `jos_content` AS `c`
INNER JOIN `jos_associations` AS `a` ON `a`.`id` = `c`.`id` AND `a`.`context` = :context
INNER JOIN `jos_associations` AS `a2` ON `a`.`key` = `a2`.`key`
INNER JOIN `jos_content` AS `c2` ON `a2`.`id` = `c2`.`id`
INNER JOIN `jos_categories` AS `ca` ON `c2`.`catid` = `ca`.`id` AND `ca`.`extension` = :extension2
...
```

### Expected result AFTER applying this Pull Request

34 queries in total, see a single query per each `ArticlesModel::getItems()` call.

A new single query is like:

```
SELECT `c2`.`language`,`c`.`id` AS `pk`,CONCAT_WS(':', `c2`.`id`, `c2`.`alias`) AS `id`,CONCAT_WS(':', `ca`.`id`, `ca`.`alias`) AS `catid`
FROM `jos_content` AS `c`
INNER JOIN `jos_associations` AS `a` ON `a`.`id` = `c`.`id` AND `a`.`context` = :context
INNER JOIN `jos_associations` AS `a2` ON `a`.`key` = `a2`.`key`
INNER JOIN `jos_content` AS `c2` ON `a2`.`id` = `c2`.`id`
INNER JOIN `jos_categories` AS `ca` ON `c2`.`catid` = `ca`.`id` AND `ca`.`extension` = :extension2
...
```

### Link to documentations

No documentation changes for docs.joomla.org needed
No documentation changes for manual.joomla.org needed
